### ISSUE TITLE
[Core] Actors leaking memory 

### ISSUE THREAD

**Issue by Samreay** [ray-project/ray#24216]
### What happened + What you expected to happen

I noticed an issue in production with our ray server continually chewing up more and more memory:

<img width="1410" alt="image" src="https://user-images.githubusercontent.com/6748505/165441366-90288129-0729-446b-9069-9c87e2d94fd7.png">

 
So I went and made a small reproduction of our ray setup showing that actors are not releasing memory when they should.

### Versions / Dependencies

```
ray==1.12.0
python==3.8.5
```

### Reproduction script

The important part here is prior to the main section. The main code simply validates this behaviour in and out of `local_mode`, with different `num_cpu` values. Our case is that we have a stateful actor. We set its state once, and then use this state to do repeated evaluations of a function at different coordinates. However, we do not get our memory back after removing this state.

```python
from collections import defaultdict
import os
import logging


import psutil
import ray
import numpy as np
import matplotlib.pyplot as plt

logging.basicConfig(level=logging.INFO)


@ray.remote
class Actor:
    def __init__(self):
        self.something_to_store = None

    def set_something(self, something_to_store):
        self.something_to_store = something_to_store

    def compute(self):
        return 1


def simulate_call(workers, pass_object):
    something_to_store = {"data": np.random.random((100000, 2)).tolist()}
    if pass_object:
        ray.get([w.set_something.remote(something_to_store) for w in workers])
    results = ray.get([w.compute.remote() for w in workers])
    return results


if __name__ == "__main__":

    print(f"Starting memory test with versio {ray.__version__}")
    results = defaultdict(list)
    process = psutil.Process(os.getpid())
    num_samples = 100
    scale = 1024**2
    
    # This is just iterating through parameters so I can check the memory issue is consistent
    for num_cpu in [1, 2, 4]:
        for local_mode in [True, False]:
            if ray.is_initialized():
                ray.shutdown()
            ray.init(num_cpus=num_cpu, num_gpus=0, include_dashboard=False, local_mode=local_mode)
            for pass_object in [True, False]:
                workers = [Actor.remote() for _ in range(num_cpu)]
                baseline = process.memory_info().rss / scale
                key = f"{pass_object=} {num_cpu=} {local_mode=}"
                print(key)
                while len(results[key]) < num_samples:
                    simulate_call(workers, pass_object)
                    results[key].append(process.memory_info().rss / scale - baseline)

    # Plotting out the memory usage as a function of iteration
    fig, axes = plt.subplots(nrows=2, sharex=True)
    for k, v in results.items():
        p = "pass_object=True" in k
        l = "local_mode=True" in k
        ax = axes[0] if l else axes[1]
        ax.plot(v, label=k, ls="--" if p else "-")
        ax.text(num_samples * 1.05, v[-1], k)
    axes[1].set_xlabel("Iteration")
    for ax in axes:
        ax.set_ylabel("Delta Memory usage (MB)")
        ax.set_xscale("log")
    fig.savefig("memory_usage.png", dpi=300, bbox_inches="tight")
```

Running this script produces this output on my machine. When `pass_obejct=False`, we have no memory issues at all, because we never set the actor's state. Then, we can see that the memory used by ray continually increases when `local_mode=True`,and it appears the state sent to the actors is never freed, even when the reference to `something_to_store` is updated and the old reference is lost. (Note this increase in memory is linear with iteration, I have the scale set to log). When `local_mode=False`, the state is *sometimes* freed, but still shows growing memory over time, when I would expect it to hold a single reference to the state at a given time, and then when this is freed, for the memory usage to return to the `pass_object=False` baseline.

Local mode results shown on top. Normal mode shown on bottom.

![memory_usage](https://user-images.githubusercontent.com/6748505/165297609-4562ea71-5393-4b17-9e07-bc6f4d97e6d6.png)

I have also tried:

* Explicitly calling `del` on the `something_to_store` objects, both in the actors and outside.
* Implementing a shutdown method on the actors to call after gathering the results to both `del` and set the state to None
* Explicitly calling `ray.kill` on the actors

However none of these free the memory that was originally allocated.

### Issue Severity

High: It blocks me from completing my task.

**Comment by Samreay** [comment]
To aid in reproducing this, I containerised and ran `memray` to produce this flamechart.

`replicate.py`:

```python
import ray
import numpy as np


@ray.remote
class Actor:
    def __init__(self):
        self.something_to_store = None

    def set_something(self, something_to_store):
        self.something_to_store = something_to_store

    def compute(self):
        return 1


def simulate_call(workers):
    something_to_store = {"data": np.random.random((100000, 2)).tolist()}
    ray.get([w.set_something.remote(something_to_store) for w in workers])
    results = ray.get([w.compute.remote() for w in workers])
    return results


if __name__ == "__main__":
    print(f"Starting memory test with versio {ray.__version__}")
    num_cpu = 6
    ray.init(num_cpus=num_cpu, num_gpus=0, include_dashboard=False, local_mode=True)
    workers = [Actor.remote() for _ in range(num_cpu)]
    for _ in range(100):
        simulate_call(workers)
```

Dockerfile:

```dockerfile
FROM python:3.8-slim AS builder

RUN pip install -U pip && pip install ray numpy memray
COPY . /
CMD memray run --native -o output.bin replicate.py && memray flamegraph -o flame.html output.bin
```

Zip of the flamechart attached, and here's a screenshot of the memory usage:

<img width="1114" alt="image" src="https://user-images.githubusercontent.com/6748505/165441132-039db39f-eefa-464f-8854-a272e7213484.png">

The main memory usage comes from:

```
ray::core::CoreWorkerProcess::Initialize(ray::core::CoreWorkerOptions const&) at <unknown>:0
ray::core::CoreWorkerProcessImpl::CoreWorkerProcessImpl(ray::core::CoreWorkerOptions const&) at <unknown>:0
ray::stats::internal::Stats::Stats(std::string const&, std::string const&, std::vector<std::string, std::allocator<std::string> >, std::vector<double, std::allocator<double> >, std::function<void (std::string const&, std::string, std::vector<opencensus::tags::TagKey, std::allocator<opencensus::tags::TagKey> >, std::vector<double, std::allocator<double> > const&)>)::{lambda()#1}::operator()() const at <unknown>:0
void ray::stats::internal::RegisterView<(ray::stats::StatsType)2>(std::string const&, std::string const&, std::vector<opencensus::tags::TagKey, std::allocator<opencensus::tags::TagKey> > const&, std::vector<double, std::allocator<double> > const&) [clone .isra.0] at <unknown>:0
ray::stats::internal::RegisterAsView(opencensus::stats::ViewDescriptor, std::vector<opencensus::tags::TagKey, std::allocator<opencensus::tags::TagKey> > const&) at <unknown>:0
opencensus::stats::StatsExporterImpl::AddView(opencensus::stats::ViewDescriptor const&) at <unknown>:0
std::__detail::_Map_base<std::string, std::pair<std::string const, std::unique_ptr<opencensus::stats::View, std::default_delete<opencensus::stats::View> > >, std::allocator<std::pair<std::string const, std::unique_ptr<opencensus::stats::View, std::default_delete<opencensus::stats::View> > > >, std::__detail::_Select1st, std::equal_to<std::string>, std::hash<std::string>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true>, true>::operator[](std::string const&) at <unknown>:0
operator new(unsigned long) at <unknown>:0
```

[flame3.html.gz](https://github.com/ray-project/ray/files/8568817/flame3.html.gz)

**Comment by Samreay** [comment]
For completeness, here is the flamegraph when the above is run, but with `local_model=False` to reflect production conditions.

Primary memory usage from:

```
ray::core::CoreWorkerProcessImpl::CoreWorkerProcessImpl(ray::core::CoreWorkerOptions const&) at <unknown>:0
ray::stats::internal::Stats::Stats(std::string const&, std::string const&, std::vector<std::string, std::allocator<std::string> >, std::vector<double, std::allocator<double> >, std::function<void (std::string const&, std::string, std::vector<opencensus::tags::TagKey, std::allocator<opencensus::tags::TagKey> >, std::vector<double, std::allocator<double> > const&)>)::{lambda()#1}::operator()() const at <unknown>:0
void ray::stats::internal::RegisterView<(ray::stats::StatsType)2>(std::string const&, std::string const&, std::vector<opencensus::tags::TagKey, std::allocator<opencensus::tags::TagKey> > const&, std::vector<double, std::allocator<double> > const&) [clone .isra.0] at <unknown>:0
ray::stats::internal::RegisterAsView(opencensus::stats::ViewDescriptor, std::vector<opencensus::tags::TagKey, std::allocator<opencensus::tags::TagKey> > const&) at <unknown>:0
opencensus::stats::StatsExporterImpl::AddView(opencensus::stats::ViewDescriptor const&) at <unknown>:0
std::__detail::_Map_base<std::string, std::pair<std::string const, std::unique_ptr<opencensus::stats::View, std::default_delete<opencensus::stats::View> > >, std::allocator<std::pair<std::string const, std::unique_ptr<opencensus::stats::View, std::default_delete<opencensus::stats::View> > > >, std::__detail::_Select1st, std::equal_to<std::string>, std::hash<std::string>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true>, true>::operator[](std::string const&) at <unknown>:0
operator new(unsigned long) at <unknown>:0
```

[flame4.html.gz](https://github.com/ray-project/ray/files/8568911/flame4.html.gz)

**Comment by rkooo567** [comment]
Hey @Samreay I have had no bandwidth to take a look at this issue, but I will take it over shortly. Please remind me if I don't update the status in 2 weeks.

**Comment by Samreay** [comment]
Hey @rkooo567, pinging you as requested :)

**Comment by rkooo567** [comment]
Thank you! It is currently tracked, but I have no capacity to take a look at the issue in another 2 weeks...

**Comment by mfojtak** [comment]
Also having the same issue. In my case it is even more severe as my process allocates GPU memory. Therefore I can run it only once until GPU gets swamped.

**Comment by rkooo567** [comment]
Hey @scv119 I think we should finish this before Ray 2.0 release. Can you add a 2.0 blocker flag if you agree?

**Comment by clarng** [comment]
@mfojtak  do you have more context of the issue you are seeing? Wanted to make sure they are the same issue and not a separate one for investigation

**Comment by afero** [comment]
Hi there,
We are seeing a very similar issue. After running `ray.tune` jobs, the memory on the head node never goes back down to what it was before the job was run. We are running ray tune on a kuberay cluster. The helm chart is adapted from https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray-cluster.complete.yaml. Let me know if the helm chart we’re using would be helpful. We were able to reproduce this with a Ray tune example (below).

<img width="756" alt="image" src="https://user-images.githubusercontent.com/9958109/174128402-36c89f50-568b-4707-ba20-d2eda549240c.png">

We've noticed that there are idle workers that remain on the head node after the HPO tune jobs. If we kill those processes manually, the memory will go down but not back to the same level as before running the job.
<img width="1142" alt="image" src="https://user-images.githubusercontent.com/9958109/174128600-4d7de372-e5d1-49bc-ae2a-9c70a065e37f.png">
 

To reproduce:

```
import sklearn.datasets
import sklearn.metrics
import sklearn.datasets
import sklearn.metrics
import xgboost as xgb

from ray import tune
from sklearn.model_selection import train_test_split

import ray


def train_breast_cancer(config):
     # Load dataset
     data, labels = sklearn.datasets.load_breast_cancer(return_X_y=True)
     # Split into train and test set
     train_x, test_x, train_y, test_y = train_test_split(
         data, labels, test_size=0.25)
     # Build input matrices for XGBoost
     train_set = xgb.DMatrix(train_x, label=train_y)
     test_set = xgb.DMatrix(test_x, label=test_y)
     # Train the classifier
     results = {}
     xgb.train(
         config,
         train_set,
         evals=[(test_set, "eval")],
         evals_result=results,
         verbose_eval=False)
     # Return prediction accuracy
     accuracy = 1. - results["eval"]["error"][-1]
     tune.report(mean_accuracy=accuracy, done=True)


if __name__ == "__main__":
    ray.init(address="ray://127.0.0.1:10001")
    
    config = {
        "objective": "binary:logistic",
        "eval_metric": ["logloss", "error"],
        "max_depth": tune.randint(1, 9),
        "min_child_weight": tune.choice([1, 2, 3]),
        "subsample": tune.uniform(0.5, 1.0),
        "eta": tune.loguniform(1e-4, 1e-1)
    }
    analysis = tune.run(
        train_breast_cancer,
        resources_per_trial={"cpu": 1},
        config=config,
        num_samples=10)
```

**Comment by rkooo567** [comment]
cc @clarng can you let us know what's the current status of this issue?

**Comment by scv119** [comment]
@clarng should we summarize and close this one?

**Comment by clarng** [comment]
Will send a summary, thanks

**Comment by nollied** [comment]
i'm having basically the same exact issue. btw this is a well put together issue @Samreay kudos

**Comment by clarng** [comment]
It seems there are several issues reported here, which we will comment on individually.
1 Memory leak when local_mode is enabled
2 Memory leak when local_mode is disabled
3 Tune not freeing memory when job completes

* Memory leak when local_mode is enabled
Local mode is an experimental feature that is no longer maintained and will be removed in the future. For debugging consider using the Ray debugger. We have added a deprecation warning when this is enabled.

* Memory leak when local_mode is disabled
I have done several memory profiles. There seems to be a bigger memory bump on Mac vs on Linux. In the Linux case the initial bump looks more reasonable and is related to the data structure it creates to manage the driver / actor and communication between components : grpc, service factory
[grpc_profile.pdf](https://github.com/ray-project/ray/files/9201208/grpc_profile.pdf)
[service_factory_profile.pdf](https://github.com/ray-project/ray/files/9201209/service_factory_profile.pdf)
Script used: https://gist.github.com/clarng/831be12849f115c9b997b1e3745e15d8
RSS / USS dumps from the script : https://gist.github.com/clarng/f91f35bded6d6c56dec67efa861cb1a0

When i strip the beginning part of the script from the profile and only measure the steady state delta these go away
[steady_state.pdf](https://github.com/ray-project/ray/files/9201216/steady_state.pdf)

* Tune not freeing memory when job completes
Filed a separate issue to track this


Would love to know if there are things not mentioned or covered by the update - it would be best to create a new issue and track the discussion separately. Thanks for filing this issue!

**Comment by nollied** [comment]
> Local mode is an experimental feature that is no longer maintained and will be removed in the future. For debugging consider using the Ray debugger. We have added a deprecation warning when this is enabled.
@clarng why are you removing local mode? it's much more convenient to develop on a serial thread than to spin up all of the ones available on my machine.

> When i strip the beginning part of the script from the profile and only measure the steady state delta these go away
what do you mean?

**Comment by pablogsal** [comment]
👋 Hi @Samreay,

I am one of the authors of `memray`. We are collecting success stories [here](https://github.com/bloomberg/memray/discussions/226). If you have a minute, do you mind leaving a short message on how `memray` help with this issue? Knowing how we managed to help will help us track trends and target areas for improvement, prioritize new features and development and identify potential bugs or areas of confusion.

Thanks a lot for your consideration and for helping us improve the profiler :)
